### PR TITLE
app_nonat test

### PR DESCRIPTION
**Pull request by sadov**
At this time blocked by https://github.com/lf-edge/eden/issues/589.

Signed-off-by: Oleg Sadov <oleg@zededa.com>

### CODE REVIEW

**Comment by sadov** [comment]
The test fails in the current state. EVE configs form this test:
```
"apps": [
        {
"interfaces": [
...
        {
    "name": "default",
    "networkId": "512839bd-a655-46ad-988d-fde5da33c3bb",
    "acls": [
        {
            "matches": [    
                {               
                    "type": "ip",   
                    "value": "0.0.0.0/0"
                }               
            ],
            "id": 1         
        }
    ]
}
]
...
"networkInstances": [
...
{
    "uuidandversion": {
        "uuid": "512839bd-a655-46ad-988d-fde5da33c3bb",
        "version": "1"
    },
    "displayname": "direct",
    "instType": 1,
    "activate": true,
    "port": {
        "type": 1,
        "name": "eth0"
    },
    "cfg": {},
    "ipType": 1,
    "ip": {}
}
...
```
 and ztest:
 ```
"apps": [
        {
"interfaces": [
...
{
    "name": "direct",
    "networkId": "72a14885-73c4-4fe7-a4b2-f919c05bf065",
    "addr": "",
    "hostname": "",
    "cryptoEid": "",
    "lispsignature": "",
    "pemcert": null,
    "pemprivatekey": null,
    "macAddress": "",
    "acls": [
        {
            "matches": [
                {
                    "type": "ip",
                    "value": "0.0.0.0/0"
                }
            ],
            "actions": [],
            "name": "",
            "id": 3,
            "dir": "BOTH"
        }
    ]
}
]
...
"networkInstances": [
...
{
    "uuidandversion": {
        "uuid": "72a14885-73c4-4fe7-a4b2-f919c05bf065",
        "version": ""
    },
    "displayname": "TestAppNonat-sc-supermicro-e100-8-NI-2021-04-14T02-00-25",
    "instType": "ZnetInstSwitch",
    "activate": true,
    "port": {
        "type": "PhyIoNoop",
        "name": "eth0"
    },
    "cfg": {
        "oconfig": "",
        "lispConfig": null,
        "type": "ZNetOConfigVPN"
    },
    "ipType": "IPV4", 
    "ip": null,
    "dns": []
}
...
```
 looks more or less similar, but pings through eth1 not passed.
 
 Log file:
[eden_test-app_nonat.015.log](https://github.com/lf-edge/eden/files/6375711/eden_test-app_nonat.015.log)

**Comment by sadov** [comment]
After adding the default gw route for eth1 to the test, the ping passes. But it still gets through after removing the acl rules from the `direct` network.

**Comment by zed-sadov** [comment]
With a such changes on EVE:
https://github.com/lf-edge/eve/pull/2042
test passed:
```
eden test tests/eclient/ -e app_nonat -v debug
```

File: tests/eclient/testdata/app_nonat.txt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+message 'Testing of network'
+exec sleep 20
+exec -t 1m bash ping.sh
+stdout '0% packet loss'
+
+message 'Switching network'
+eden pod modify eclient --networks indirect --networks=direct --acl='direct:'
+test eden.app.test -test.v -timewait 30m RUNNING eclient
+
```
Comment (giggsoff) [inline tests/eclient/testdata/app_nonat.txt:43]: @sadov please add here
```
message 'Checking accessibility'
exec -t 5m bash wait_ssh.sh
```
`eden pod modify` will restart app in this case, so we must wait for ssh

File: tests/eclient/testdata/app_nonat.txt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+! exec -t 1m bash ping.sh
+stdout '100% packet loss'
+
+message 'Resource cleanng'
```
Comment (giggsoff) [inline tests/eclient/testdata/app_nonat.txt:47]: cleanng->cleaning

File: tests/eclient/testdata/app_nonat.txt [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+eden network ls
+! stdout '^direct\s'
+! stdout '^indirect\s'
+
+-- wait_ssh.sh --
+EDEN={{EdenConfig "eden.root"}}/{{EdenConfig "eden.bin-dist"}}/{{EdenConfig "eden.eden-bin"}}
+HOST=$($EDEN eve ip)
+
+for p in $*
```
Comment (giggsoff) [inline tests/eclient/testdata/app_nonat.txt:69]: Any reason to use this loop here?

File: tests/eclient/testdata/app_nonat.txt [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+message 'Checking accessibility'
+exec -t 5m bash wait_ssh.sh
+
+message 'Testing of network'
+exec sleep 20
+exec -t 1m bash ping.sh
+stdout '0% packet loss'
+
+message 'Switching network'
```
Comment (giggsoff) [inline tests/eclient/testdata/app_nonat.txt:40]: Please add comment, that we remove all ACLs for direct network for which one eth1 of eclient connected.

File: tests/eclient/testdata/app_nonat.txt [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    {{template "ssh"}}$HOST grep -q Ubuntu /etc/issue && break
+  done
+done
+
+-- ping.sh --
+EDEN={{EdenConfig "eden.root"}}/{{EdenConfig "eden.bin-dist"}}/{{EdenConfig "eden.eden-bin"}}
+HOST=$($EDEN eve ip)
+
+#echo {{template "ssh"}}$HOST route add default gw 192.168.0.2 dev eth1
+#{{template "ssh"}}$HOST route add default gw 192.168.0.2 dev eth1
```
Comment (giggsoff) [inline tests/eclient/testdata/app_nonat.txt:85]: Should we remove comments here?

**Review by giggsoff — CHANGES_REQUESTED** [review]
Please add wait for ssh after pod modify and take a look at my comments

File: tests/eclient/testdata/app_nonat.txt (hunk 3 above)
Comment (sadov) [reply tests/eclient/testdata/app_nonat.txt:69]: Sure -- mentioning the ACL here will make sense, but the exact wording will need to be coordinated with the network documentation and the second is planned to be created.

File: tests/eclient/testdata/app_nonat.txt (hunk 2 above)
Comment (sadov) [reply tests/eclient/testdata/app_nonat.txt:47]: Sure -- mentioning the ACL here will make sense, but the exact wording will need to be coordinated with the network documentation and the second is planned to be created.

File: tests/eclient/testdata/app_nonat.txt (hunk 1 above)
Comment (sadov) [reply tests/eclient/testdata/app_nonat.txt:43]: Reasonably

File: tests/eclient/testdata/app_nonat.txt (hunk 2 above)
Comment (sadov) [reply tests/eclient/testdata/app_nonat.txt:47]: OK

File: tests/eclient/testdata/app_nonat.txt (hunk 4 above)
Comment (sadov) [reply tests/eclient/testdata/app_nonat.txt:40]: Sure -- mentioning the ACL here will make sense, but the exact wording will need to be coordinated with the network documentation and the second is planned to be created.

File: tests/eclient/testdata/app_nonat.txt (hunk 5 above)
Comment (sadov) [reply tests/eclient/testdata/app_nonat.txt:85]: Sure

File: tests/eclient/testdata/app_nonat.txt (hunk 3 above)
Comment (giggsoff) [reply tests/eclient/testdata/app_nonat.txt:69]: We have the inner loop here for that (`for i in seq 20`), right?

File: tests/eclient/testdata/app_nonat.txt (hunk 4 above)
Comment (giggsoff) [reply tests/eclient/testdata/app_nonat.txt:40]: Well, I think if we remove all ACLs, we can be independent from spelling, we just remove them.

**Comment by giggsoff** [comment]
GCP test [failed](https://github.com/itmo-eve/eden/actions/runs/834837825) again with changes in your PR. Seems, we cannot use switch with uplink on GCP, it allocates only one address per VM. [Alias IPs](https://cloud.google.com/vpc/docs/alias-ip#alias_ip_ranges_key_properties) only usable with manual ip setting, it is not useful in our case.
So, I suggest to move the test onto large-only branch.

**Comment by sadov** [comment]
OK - moved to "large".

**Pull request merged**